arch/ia32/jmp.S

```diff
@@ -5,127 +5,145 @@
  *
  * setjmp, longjmp
  *
- * Copyright 2018-2019 Phoenix Systems
- * Author: Jan Sikorski, Andrzej Glowinski
+ * Copyright 2018-2019, 2023 Phoenix Systems
+ * Author: Jan Sikorski, Andrzej Glowinski, Hubert Badocha
  *
  * This file is part of Phoenix-RTOS.
  *
  * %LICENSE%
  */
 
+ /* jmp_buf layout:
+ * esi
+ * ebx
+ * edi
+ * ebp
+ * esp + 4
+ * return address
+ * signal mask indicator (0 - not saved, otherwise saved(GOT addr on PIC))
+ * signal mask (if saved)
+ */
+
 #define __ASSEMBLY__
 
 
 .text
 
-.globl _setjmp
-.type _setjmp, %function
-_setjmp:
+
+.macro __SETJMP
 	/* Store registers into jmpbuf */
-	movl 4(%esp),   %edx
-	movl   %ebx,   (%edx)
-	movl   %esi,  4(%edx)
-	movl   %edi,  8(%edx)
-	movl   %ebp, 12(%edx)
+	leal 4(%esp),    %ecx
+	movl  (%ecx),    %edx
+	movl   %esi,    (%edx)
+	movl   %ebx,   4(%edx)
+	movl   %edi,   8(%edx)
+	movl   %ebp,  12(%edx)
 	/* Store stack pointer pointing before return address */
-	leal 4(%esp),   %ecx
-	movl   %ecx, 16(%edx)
-	/* Store jump address */
-	leal  (1f),     %ecx
-	movl   %ecx, 20(%edx)
+	movl   %ecx,  16(%edx)
 	/* Store return address */
-	movl  (%esp),   %ecx
-	movl  %ecx,  24(%edx)
-	/* Store result */
-	movl  $0,       %eax
-	ret
-1:
-	/* Jump point. Push return address to stack */
-	pushl  24(%edx)
+	movl  (%esp),    %ecx
+	movl   %ecx,  20(%edx)
+.endm
+
+
+.globl _setjmp
+.type _setjmp, %function
+_setjmp:
+	__SETJMP
+	/* Set no mask indicator */
+	movl   $0,    24(%edx)
+	/* Store result (0) */
+	xorl   %eax,     %eax
 	ret
 .size _setjmp, .-_setjmp
 
 
 .globl setjmp
 .type setjmp, %function
 setjmp:
-        /* Get signal mask */
-	pushl  $0
-	pushl  $0
-	call   signalMask
-	addl   $8,      %esp
-	/* Store registers into jmpbuf */
-	movl 4(%esp),   %edx
-	movl   %ebx,   (%edx)
-	movl   %esi,  4(%edx)
-	movl   %edi,  8(%edx)
-	movl   %ebp, 12(%edx)
-	/* Store stack pointer pointing before return address */
-	leal 4(%esp),   %ecx
-	movl   %ecx, 16(%edx)
-	/* Store jump address */
-	leal  (1f),     %ecx
-	movl   %ecx, 20(%edx)
-	/* Store return address */
-	movl  (%esp),   %ecx
-	movl   %ecx, 24(%edx)
-	/* Store signal mask */
-	movl   %eax, 28(%edx)
-	/* Store result */
-	movl   $0,      %eax
-	ret
-1:
-	/* Jump point. Push return address to stack */
-	pushl 24(%edx)
-	/* Store longjmp return value */
-	pushl    %eax
-	/* Restore signal mask */
-	movl  28(%edx), %ecx
-	pushl    $0xffffffff
-	pushl    %ecx
+	__SETJMP
+    /* Get signal mask */
+#if __pic__ == 0
+	pushl    $0
+	pushl    $0
 	call     signalMask
-	addl     $8,    %esp
-	/* Restore longjmp return value */
-	popl     %eax
+	addl     $8,      %esp
+	/* Store signal mask indicator */
+	movl     $1,   24(%edx)
```

Comment:
What value will the `edx` register have?
Same if `__pic__ != 0`.

Reply:
unchanged, as signalMask is syscall and in syscalls only eax is modified. I'll add a comment